### PR TITLE
SY-3826: For Loop Support in Arc

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-3826](https://linear.app/synnax/issue/SY-3826/improve-series-capabilities-in-arc)

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR implements comprehensive for-loop support in the Arc language, adding multiple loop forms with full compilation to WebAssembly.

## Key Changes

- **Grammar & Parser**: Added `for`, `break`, and `continue` keywords with support for 4 loop forms: range loops (`for i := range(n)`), series iteration (`for i, x := series`), conditional loops (`for condition`), and infinite loops (`for`)
- **Range Loops**: Support 1-3 arguments with bidirectional iteration (negative steps), runtime direction detection for optimal exit conditions
- **Compiler**: Implemented WASM code generation with proper block nesting, label depth tracking, and break/continue support using `LoopStack` and `LoopDepth` context fields
- **Analyzer**: Added type checking for range arguments (must be integers), series iteration validation, loop variable immutability enforcement, and scope-based break/continue validation
- **Symbol Table**: Introduced `KindLoop` and `KindLoopVariable` for loop scopes and immutable iteration variables
- **LSP Support**: Added completions, semantic highlighting, and code folding for loop constructs
- **Testing**: Excellent test coverage across all layers - parser tests (syntax validation), analyzer tests (semantic validation), compiler tests (WASM bytecode verification), and runtime integration tests (execution verification with edge cases)

## Test Coverage Highlights

- Range loops tested with all argument combinations including edge cases: empty ranges, reversed bounds, negative steps, and step boundary conditions
- Series iteration tested with single and dual-variable forms
- Nested loops tested up to 3 levels deep with proper label resolution
- Break/continue tested in various contexts including nested if statements
- Stateful variable accumulation across reactive executions verified

<h3>Confidence Score: 5/5</h3>

- This PR is safe to merge with high confidence - comprehensive implementation with excellent test coverage across all layers
- The implementation is thorough and well-tested. All changes follow established patterns in the codebase (similar to if-statement implementation). The PR includes parser tests, analyzer tests with negative cases, compiler tests with WASM bytecode verification, and extensive runtime integration tests covering edge cases. The code handles complex scenarios like bidirectional range iteration, nested loops with correct label depth calculation, and proper scope management for break/continue statements.
- No files require special attention

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| arc/go/parser/ArcParser.g4 | Added grammar rules for `for`, `break`, and `continue` statements with support for multiple loop forms (range, series iteration, conditional, infinite) |
| arc/go/compiler/statement/loop.go | Implemented comprehensive for-loop compilation with support for range loops (1-3 args with bidirectional step), series iteration (single/dual-variable), conditional, and infinite loops with proper WASM label management for break/continue |
| arc/go/analyzer/statement/statement.go | Added semantic analysis for for-loops including type checking for range arguments, series iteration validation, loop variable scope management, and break/continue validation |
| arc/go/symbol/symbol.go | Added `KindLoop` and `KindLoopVariable` symbol kinds for loop scope tracking and immutable iteration variables |
| arc/go/compiler/context/context.go | Added `LoopEntry`, `LoopDepth`, and `LoopStack` for tracking nested loop depth and computing correct WASM branch labels for break/continue |
| arc/go/compiler/statement/statement_test.go | Added comprehensive compiler tests covering range loops (1-3 args), conditional loops, infinite loops, break/continue, and nested loops with WASM bytecode verification |
| arc/go/analyzer/statement/statement_test.go | Added analyzer tests for all loop forms including range validation, series iteration type checking, break/continue scope validation, and loop variable immutability |
| arc/go/runtime/wasm/wasm_test.go | Added extensive runtime integration tests covering all loop forms including edge cases (negative steps, empty ranges, nested loops) with actual execution verification |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    Start[For Loop Syntax] --> Parse[Parser: ArcParser.g4]
    Parse --> |4 Forms| ForClause{For Clause Type}
    
    ForClause --> |i := range n| RangeLoop[Range Loop]
    ForClause --> |i, x := series| SeriesLoop[Series Iteration]
    ForClause --> |condition| CondLoop[Conditional Loop]
    ForClause --> |empty| InfLoop[Infinite Loop]
    
    RangeLoop --> |1 arg: end| Range1[0 to end, step=1]
    RangeLoop --> |2 args: start,end| Range2[start to end, step=1]
    RangeLoop --> |3 args: start,end,step| Range3[start to end, step=step]
    
    Range1 --> Analyze[Analyzer: Type Checking]
    Range2 --> Analyze
    Range3 --> Analyze
    SeriesLoop --> Analyze
    CondLoop --> Analyze
    InfLoop --> Analyze
    
    Analyze --> |Create Loop Scope| Symbols[Symbol Table]
    Symbols --> |KindLoop, KindLoopVariable| Compile[Compiler: WASM Generation]
    
    Compile --> |Range| CompileRange[compileForRange]
    Compile --> |Series| CompileSeries[compileForSeriesIteration]
    Compile --> |Condition| CompileCond[compileForCondition]
    Compile --> |Infinite| CompileInf[compileForInfinite]
    
    CompileRange --> WASM[WASM Bytecode]
    CompileSeries --> WASM
    CompileCond --> WASM
    CompileInf --> WASM
    
    WASM --> |block $break| Block1[Outer Block]
    Block1 --> |loop $continue| Loop1[Loop Header]
    Loop1 --> |Condition Check| CondCheck{Exit?}
    CondCheck --> |Yes| Break[br to $break]
    CondCheck --> |No| Body[Loop Body]
    Body --> |break stmt| BreakStmt[br to $break]
    Body --> |continue stmt| ContStmt[br to $continue]
    Body --> Increment[Increment/Next]
    Increment --> |br 0| Loop1
    
    Break --> End[End Block]
    ContStmt --> Increment
```

<sub>Last reviewed commit: 3c34b6f</sub>

### CODE REVIEW

File: docs/site/src/pages/reference/control/arc/concepts/stateful-variables.mdx [hunk 1]
```diff
@@ -114,22 +114,26 @@ func toggle(trigger u8) u8 {
...
+**[`for` loops](/reference/control/arc/reference/loops)** process data you already have.
+If a function receives a series, a loop can walk through it in a single execution:
 
-```python
-total = 0
-for reading in readings:
-    total += reading
-average = total / len(readings)
+```arc
+func batch_average(data series f64) f64 {
```
Comment (pjdotson) [inline docs/site/src/pages/reference/control/arc/concepts/stateful-variables.mdx]: do we have any test / support for looping through the series? Unless I'm being retarded
(always a possibility) I don't see anything.

File: docs/site/src/pages/reference/control/arc/reference/loops.mdx [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        if x > peak {
+            peak = x
+            peak_idx = i
+        }
+    }
+    return peak_idx
+}
+```
+
+The index starts at 0 and has type `i32`.
```
Comment (pjdotson) [inline docs/site/src/pages/reference/control/arc/reference/loops.mdx:64]: should it be `i32` or `i64` or a uint type?

File: docs/site/src/pages/reference/control/arc/concepts/reactive-execution.mdx [hunk 3]
```diff
@@ -48,8 +48,8 @@ You declare how data flows through your system, and Arc handles the scheduling:
 tank_pressure -> scale{factor=2.0} -> pressure_display
 ```
 
-No loops. No polling. When `tank_pressure` receives new data, Arc automatically runs
-`scale` and updates `pressure_display`.
+No polling. No manual scheduling. When `tank_pressure` receives new data, Arc
+automatically runs `scale` and updates `pressure_display`.
```
Comment (pjdotson) [inline docs/site/src/pages/reference/control/arc/concepts/reactive-execution.mdx]: a lot of the language on this page (specifically these super short sentences) read a bit
too much like GPT marketing slop

**Review by pjdotson — CHANGES_REQUESTED** [review]

File: docs/site/src/pages/reference/control/arc/concepts/stateful-variables.mdx (hunk 1 above)
Comment (sy-nico) [reply docs/site/src/pages/reference/control/arc/concepts/stateful-variables.mdx]: None. But more importantly, this functionality does not exist at the moment.

Comment (sy-nico) [reply docs/site/src/pages/reference/control/arc/concepts/stateful-variables.mdx]: This functionality will not be introduced in this PR. Unless requested, only the built-in nodes will be able to handle series operation, users will not be able to define functions that take a series input. Docs have been updated.

File: docs/site/src/pages/reference/control/arc/reference/loops.mdx (hunk 2 above)
Comment (sy-nico) [reply docs/site/src/pages/reference/control/arc/reference/loops.mdx:64]: It's a WASM thing. The series host functions (`series_len`, `series_get_element_*`, etc.) all use `i32` for handles and indices because that's WASM32's native addressing type. Making the loop index `i64` would just mean narrowing/widening at every host call for no benefit.

Hardcoded at `statement.go:423` and the hidden loop vars at lines 561-563.
